src/ring_discord_auth/validation.clj

```diff
@@ -0,0 +1,181 @@
+               ^CharsetEncoder encoder (try (.newEncoder cs) (catch UnsupportedOperationException _ nil))]
+              (when (.canEncode encoder str)
+                (read-all-bytes (.encode encoder (CharBuffer/wrap str))))))
+
+(defn hex->bytes
+  "Converts the given string representing a hexadecimal number to a byte array.
+  Each byte in the resulting array comes from 2 digits in the string.
+
+  If the string cannot be converted, returns `nil`"
+  ^bytes [^String hex-str]
+  (let [len (count hex-str)
+        result (byte-array (quot (inc len) 2))]
+    (try
+      (doseq [[i hex-part] (map-indexed vector (map (partial apply str) (partition-all 2 hex-str)))]
+        (aset result i (unchecked-byte (Short/parseShort hex-part 16))))
+      result
+      (catch NumberFormatException _ nil))))
+
+(defn bytes->hex
+  "Convert byte array to hex string."
+  [^bytes byte-array]
+  (let [hex [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f]]
+    (letfn [(hexify-byte [b]
+              (let [v (bit-and b 0xFF)]
+                [(hex (bit-shift-right v 4)) (hex (bit-and v 0x0F))]))]
+      (str/join (mapcat hexify-byte byte-array)))))
+
+(defn generate-keypair
+  "Generate Ed25519 key pair.
+
+  Return {:private `Ed25519PrivateKeyParameters`
+          :public `Ed25519PublicKeyParameters`}"
+  []
+  (let [random (SecureRandom.)
+        kpg    (Ed25519KeyPairGenerator.)
+        _ (.init kpg (Ed25519KeyGenerationParameters. random))
+        key-pair (.generateKeyPair kpg)]
+    {:private (cast Ed25519PrivateKeyParameters (.getPrivate key-pair))
+     :public  (cast Ed25519PublicKeyParameters (.getPublic key-pair))}))
+
+(defn new-signer
+  "Return new instance of `Ed25519Signer` initialized by private key"
+  [private-key]
+  (let [signer (Ed25519Signer.)]
+    (.init signer true private-key)
+    signer))
+
+(defn sign
+  "Generate signature for msg byte array.
+
+  Return byte array with signature."
+  [^Ed25519Signer signer msg-bytes]
+  (.update signer msg-bytes 0 (alength msg-bytes))
+  (.generateSignature signer))
+
+(defn new-verifier
+  "Return new instance of `Ed25519Signer` initialized by public key."
+  [public-key]
+  (let [signer (Ed25519Signer.)]
+    (.init signer false public-key)
+    signer))
+
+(defn verify
+  "Verify signature for msg byte array.
+
+  Returns `true` if valid signature and `false` if not."
+  [^Ed25519Signer signer msg-bytes signature]
+  (.update signer msg-bytes 0 (alength msg-bytes))
+  (.verifySignature signer signature))
+
+(defn verify-request
+  "Verify discord payload with app public-key, request body, signature and timestamp headers
+
+  Returns `true` if the message is authentic, `false` if not."
+  [public-key timestamp body signature]
+  (verify (new-verifier (Ed25519PublicKeyParameters. (hex->bytes public-key) 0))
+          (.getBytes (str timestamp body) "utf8")
+          (hex->bytes signature)))
```

Comment:
Could you remove these two functions and instead add support to pass an `Ed25519Signer` to `authentic?`? 
I.e. the signature of authentic? would become `[signature-bytes message-bytes public-key-bytes-or-verifier]` and same for the other arity. That would remove potential confusion as to why there are two functions that essentially do the same thing

Reply:
removed `verify-request` :+1: